blake2/src/blake2b/guts.rs

```diff
@@ -0,0 +1,565 @@
+        #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+        {
+            if let Some(sse41_impl) = Self::sse41_if_supported() {
+                return sse41_impl;
+            }
+        }
+        Self::portable()
+    }
+
+    pub fn portable() -> Self {
+        Implementation(Platform::Portable)
+    }
+
+    #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+    #[allow(unreachable_code)]
+    pub fn sse41_if_supported() -> Option<Self> {
+        // Check whether SSE4.1 support is assumed by the build.
+        #[cfg(target_feature = "sse4.1")]
+        {
+            return Some(Implementation(Platform::SSE41));
+        }
+        // Otherwise dynamically check for support if we can.
+        #[cfg(feature = "std")]
+        {
+            if is_x86_feature_detected!("sse4.1") {
+                return Some(Implementation(Platform::SSE41));
+            }
+        }
+        None
+    }
+
+    #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+    #[allow(unreachable_code)]
+    pub fn avx2_if_supported() -> Option<Self> {
+        // Check whether AVX2 support is assumed by the build.
+        #[cfg(target_feature = "avx2")]
+        {
+            return Some(Implementation(Platform::AVX2));
+        }
+        // Otherwise dynamically check for support if we can.
+        #[cfg(feature = "std")]
+        {
+            if is_x86_feature_detected!("avx2") {
+                return Some(Implementation(Platform::AVX2));
+            }
+        }
+        None
+    }
+
+    pub fn degree(&self) -> usize {
+        match self.0 {
+            #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+            Platform::AVX2 => avx2::DEGREE,
+            #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+            Platform::SSE41 => sse41::DEGREE,
+            Platform::Portable => 1,
+        }
+    }
+
+    pub fn compress1_loop(
+        &self,
+        input: &[u8],
+        words: &mut [Word; 8],
+        count: Count,
+        last_node: LastNode,
+        finalize: Finalize,
+        stride: Stride,
+    ) {
+        match self.0 {
+            #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+            Platform::AVX2 => unsafe {
+                avx2::compress1_loop(input, words, count, last_node, finalize, stride);
+            },
+            // Note that there's an SSE version of compress1 in the official C
+            // implementation, but I haven't ported it yet.
+            _ => {
+                portable::compress1_loop(input, words, count, last_node, finalize, stride);
+            }
+        }
+    }
+
+    pub fn compress2_loop(&self, jobs: &mut [Job<'_, '_>; 2], finalize: Finalize, stride: Stride) {
+        match self.0 {
+            #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+            Platform::AVX2 | Platform::SSE41 => unsafe {
+                sse41::compress2_loop(jobs, finalize, stride)
+            },
+            _ => panic!("unsupported"),
```

Comment:
@oconnor663 noticing compile failures (well, more like warnings) for this and the other cases below when testing on a `thumbv7em-none-eabi ` target (which is SOP for us).

Is there some sort of fallback on these platforms that will avoid this panic? If not, these should probably be changed to `compile_error` (and separately, we should address the issue before merging if so)

Reply:
@tarcieri are you sure you can trigger this panic in practice? If you look at how these functions are used in `compress_many`, `imp.degree()` is checked first, which should prevent this panic. In particular, I expect the `Portable` platform should get selected on `thumbv7em-none-eabi`, which should have a `degree()` of 1.